### DISCUSSION TITLE
Data communication between the extension and executed code

### DISCUSSION THREAD

**Discussion by woung717** [eclipse-theia/theia#13326] [Q&A] [answered]
I'm building a Theia extension that stores executed code results in SQLite or NeDB and displays them in the IDE. 

How should data be communicated between these components? Should the extension spawn a subprocess to manage saving results (listen results and save into the database), or can the backend contribution provide a REST API or Websocket endpoints for the code to send data upon execution? 
I found documentation suggests backend contributions support between frontend and backend communication.

**Comment by msujew** [comment, marked as answer]
Hey @woung717,

I guess it depends? Generally, it's possible to send the data to the backend first and then let the backend handle everything. See our documentation [on RPC](https://theia-ide.org/docs/json_rpc/). Having a REST API could be possible as well though, depending on whether you prefer building microservices, etc. The RPC solution would probably be recommended if you don't intend on reusing the database access.

**Reply by woung717** [reply]
Hi @msujew! Thanks for answering even on the weekend. 

> Generally, it's possible to send the data to the backend first and then let the backend handle everything.

Does this mean that backend can capture the result of executed code? I guess it is through IPC? (Assuming the code is executed by something like fork()/spawn())

**Reply by msujew** [reply]
Sure, you should be able to capture any output through IPC via the object exposed by the `fork/spawn` function of node.

**Reply by woung717** [reply]
@msujew Thank you! I understand the concepts.
There is one more question, if backend contribution expose a endpoint something like this code,
```javascript
configure(app: Application): void {
        app.get('/endpoint', (request, response) => {
            console.log(request.body);
            response.send("OK");
        });
    }
``` 
Can other processes (e.g. spawned processes) access the endpoint?

**Reply by msujew** [reply]
Yes, these work just like any other HTTP app endpoint. Unless you specifically try to secure the Express app, you can access the endpoint from any process.

**Reply by woung717** [reply]
@msujew Thanks! I appreciate your help!
